### PR TITLE
[Velux] Bugfixes (Empty list of items, HTTP/JSON, sending error, binding shutdown/removal, channel values) plus Enhancements (documentation, new channels timestamp+reload, optional inversed actuator values, support of Somfy devices)

### PR DESCRIPTION
**Pull request by gs4711**
The following changes are provided:

- Bugfix: Handle empty list of items properly.
- Bugfix: HTTP/JSON communication.
- Bugfix: aborting connection after sending error.
- Bugfix: firmware revision properly returned.
- Bugfix: close connections during binding shutdown/removal.
- Bugfix: string representation and handling of ip addresses.
- Bugfix: invalid use of myJsonBridge eliminated in case of LAN/WLAN config.
- Bugfix: initial refresh of data for item added, handling of BRIDGE_REFRESH_MSECS fixed.
- Bugfix: refresh only readable items.
- Bugfix/enhancement: adapted logging (more infos, less warnings).

- Enhancement: comments for activation with OH2.
- Enhancement: more information during actuator state modification for debugging.
- Enhancement: new channel timestamp.
- Enhancement: intensive logging for recognition of Somfy devices.
- Enhancement: optional handling for inversed actuators.
- Enhancement: eliminating usage of org.apache.commons.lang.StringUtils.isNotBlank.
- Enhancement: proper handling of unknown config items.
- Enhancement: more infos about debug settings in README.
- Enhancement: more infos in loglevel trace.
- Enhancement: provide infos about WLAN config.
- Enhancement: more detailed configuration in README added.
- Enhancement: better error handling for unknown scenes (virtual rollershutters).
- Enhancement: handling bad serial numbers (Somfy).
- Enhancement: reload action for retrieving all infos from bridge.

### CODE REVIEW

**Comment by 9037568** [comment]
If the plan is to create an OH2 binding, that should be done in the OH2 repo.

**Comment by gs4711** [comment]
> If the plan is to create an OH2 binding, that should be done in the OH2 repo.

Yes, it will be fully OH2 compliant with a separate initialization module for dealing with things.

This bug fix is necessary as OH2 calls the configuration method with a null pointer instead of an empty Dictionary.

**Comment by gs4711** [comment]
Thx, @wborn, for all the license adaptions! Now, this PR is adapted to the already approved #5843. Therefore, only minor changes have been left over.

**Comment by gs4711** [comment]
Hopefully one of the maintainers is reading this: the list of changes is mentioned/updated in the initial comment.

**Comment by kaikreuzer** [comment]
> Thx, @wborn, for all the license adaptions!

Were those somehow lost again? I see a lot of non-standard license texts and EPLv1 mentioned in them. Could you check and possibly re-run a `mvn license:format`?

**Comment by kaikreuzer** [comment]
> Yes, it will be fully OH2 compliant with a separate initialization module for dealing with things.

I am not fully clear on what that means. As it is a 1.x binding, it must not refer to Things etc., so I do nut understand why you have added things files here and removed the binding configuration.

> This bug fix is necessary as OH2 calls the configuration method with a null pointer instead of an empty Dictionary.

This PR seems to do much more that this bugfix by now...

**Comment by gs4711** [comment]
>  Could you check and possibly re-run a `mvn license:format`?

Unfortunatelly rhe re-run of a `mvn license:format` does not lead to any changes???

**Comment by gs4711** [comment]
> This PR seems to do much more that this bugfix by now...

Yes, of course: Due to the [feedback] (https://community.openhab.org/t/velux-new-openhab2-binding-feedback-welcome/32926/354) of users with OH1 and OH2, this PR contains nine bugfixes and ten extensions since 16/03/2019 in addition.

**Comment by gs4711** [comment]
> I see a lot of non-standard license texts and EPLv1

Thanks for this notice: Now, those files are manually adapted.

**Comment by kaikreuzer** [comment]
Thanks for the quick update, looks much better now.
So my remaining question is this one from above:

> As it is a 1.x binding, it must not refer to Things etc., so I do nut understand why you have added things files here and removed the binding configuration.

**Comment by gs4711** [comment]
> So my remaining question is this one from above:
> 
> > As it is a 1.x binding, it must not refer to Things etc., so I do nut understand why you have added things files here and removed the binding configuration.

The separation of the binding configuration into things files originated in the strange behaviour of paperUI which does not handle this 1.x binding properly. This part of the change (which came from the adoptions towards a full 2.x binding) could be skipped as it did not resolve this issue completely.

**Comment by kaikreuzer** [comment]
> This part of the change (which came from the adoptions towards a full 2.x binding) could be skipped

Yes, I would suggest this as Thing files must not be part of any 1.x binding and all configuration must be done on the binding level (as it is before this PR).

**Comment by gs4711** [comment]
> Yes, I would suggest this as Thing files must not be part of any 1.x binding and all configuration must be done on the binding level (as it is before this PR).

Adapted as suggested.

File: bundles/binding/org.openhab.binding.velux/ESH-INF/i18n/velux_de.properties [hunk 1]
```diff
@@ -0,0 +1,66 @@
+
+# binding
+binding.velux.name = Velux Binding  (1.x)
+binding.velux.description = Ein openHAB1 Binding f�r das Velux KLF200 Kopplungselement. Das Velux Binding kommuniziert �ber ein Velux Kopplungselement mit den unterschiedlichen Velux-Ger�ten zur Steuerung von Fenstern, Roll�den und anderen Ger�ten. Beispielsweise kann ein KLF 200 als Schnittstelle zwischen Hausleitsystemen und VELUX INTEGRA Produkten mit Funkanbindung basierend auf dem io-homecontrol-Standard eingesetzt werden.
+
+# thing types
```
Comment (kaikreuzer) [inline bundles/binding/org.openhab.binding.velux/ESH-INF/i18n/velux_de.properties]: Could you please also remove all thing and channel type translations here as well?

File: bundles/binding/org.openhab.binding.velux/README.md [hunk 2]
```diff
@@ -304,9 +428,28 @@ Notes:
 - Firmware updates are currently provided at [Velux download area](https://updates2.velux.com/).
 
 
+## How to run it under openHAB2?
```
Comment (kaikreuzer) [inline bundles/binding/org.openhab.binding.velux/README.md]: This section should be removed as it is not relevant for 1.x bindings that are made available in the official distro, see  https://www.openhab.org/docs/developer/legacy/compatibilitylayer.html#how-to-add-a-successfully-tested-1-x-add-on-to-the-distribution. If no feature is defined for this binding so far, please add it in this PR and installation will be straight forward, just like any other binding.

Comment (gs4711) [reply bundles/binding/org.openhab.binding.velux/README.md]: Thanks for those pointers.

Fixed as proposed.

File: bundles/binding/org.openhab.binding.velux/ESH-INF/i18n/velux_de.properties (hunk 1 above)
Comment (gs4711) [reply bundles/binding/org.openhab.binding.velux/ESH-INF/i18n/velux_de.properties]: Removed.

**Comment by gs4711** [comment]
Thanks, @kaikreuzer, a lot!

**Comment by gs4711** [comment]
@kaikreuzer ,

first of all, thanks for the merge of the #5833.

> I am not fully clear on what that means. As it is a 1.x binding, it must not refer to Things etc., so I do not understand why you have added things files here and removed the binding configuration.

 Sorry to bother you but after removing the ESH/things stuff the PaperUI does not provide any thing information for this compat binding. Therefore, having eliminated the bridge and thing declaration, the config-file-based configuration is the only way to handle it with PaperUI.

An ideas?

Best regards, Guenther

**Comment by kaikreuzer** [comment]
That's exactly how 1.x bindings work. Note that the main feature of 2.x bindings is to bring thing & Paper UI support, which does not exist in 1.x.

**Comment by andrewfg** [comment]
Hi Guenter @gs4711 — How are you getting on with your Velux OH v2 binding? In the meantime, I have migrated the v1 version of the NeoHub binding to v2, and also written a completely new v2 binding for SIemens smart thermostats. And now the only equipment in my home which is not on OH v2 are the Velux roof windows. So I have some skills, some motivation, and the test equipment (KLF200 and windows), to help you with the v1 —v2 migration. Do you already have a beta of the v2 binding somewhere that I can play with?

**Comment by gs4711** [comment]
Hello Andrew,

 

the oh2 binding is about to be finished in the next days (PR#2531 has already run successfully through Jenkis).

The oh2 features like discovery are not yet stable so that I will dig deeper into it before announcing it to the list.

But, in fact, with PaperUI it already looks good to me ;-)

 

Regards, Guenther

 

Von: Andrew Fiddian-Green <notifications@github.com> 
Gesendet: Donnerstag, 8. August 2019 16:38
An: openhab/openhab1-addons <openhab1-addons@noreply.github.com>
Cc: Guenther Schreiner <guenther.schreiner@smile.de>; Mention <mention@noreply.github.com>
Betreff: Re: [openhab/openhab1-addons] [Velux] Bugfixes (Empty list of items, HTTP/JSON, sending error, binding shutdown/removal, channel values) plus Enhancements (documentation, new channels timestamp+reload, optional inversed actuator values, support of Somfy d

 

Hi Guenter @gs4711 <https://github.com/gs4711>  — How are you getting on with your Velux OH v2 binding? In the meantime, I have migrated the v1 version of the NeoHub binding to v2, and also written a completely new v2 binding for SIemens smart thermostats. And now the only equipment in my home which is not on OH v2 are the Velux roof windows. So I have some skills, some motivation, and the test equipment (KLF200 and windows), to help you with the v1 —v2 migration. Do you already have a beta of the v2 binding somewhere that I can play with?

**Comment by andrewfg** [comment]
I notice that PR#4416 seems also to be an OH2 binding for Velux equipment ?? - is there some coordination with that?

**Pull request merged**